code/renderercommon/tr_image_png.c

```diff
@@ -2483,3 +2485,296 @@ void R_LoadPNG(const char *name, byte **pic, int *width, int *height)
+
+	for(h = IHDR_Height; h > 0; h--)
+	{
+		InPtr = InBuffer + (BytesPerScanline + Padding) * (h - 1);
+
+		/*
+		 *  set FilterType
+		 */
+
+		OutPtr[0] = PNG_FilterType_None;
+		OutPtr++;
+
+		for(w = 0; w < (BytesPerScanline / BytesPerPixel); w++)
+		{
+			OutPtr[0] = InPtr[0];
+			OutPtr[1] = InPtr[1];
+			OutPtr[2] = InPtr[2];
+
+			InPtr += InBytesPerPixel;
+			OutPtr += BytesPerPixel;
+		}
+	}
+
+	return(qtrue);
+}
+
+/*
+ * Write data to buffer.
+ */
+
+void WriteToBuffer(void **buffer, const void *data, size_t length)
+{
+	memcpy(*buffer, data, length);
+	*(byte**)buffer += length;
+}
+
+/*
+ * Write PNG chuck header to buffer.
+ */
+
+void WriteChunkHeader(void **buffer, void **crcPtr, PNG_ChunkCRC *CRC, int type, int length)
+{
+	struct PNG_ChunkHeader	CH;
+
+	/*
+	 *  Write chuck header
+	 */
+
+	CH.Type = BigLong(type);
+	CH.Length = BigLong(length);
+
+	WriteToBuffer(buffer, &CH, PNG_ChunkHeader_Size);
+
+	/*
+	 *  Init CRC
+	 */
+
+	*CRC = ri.zlib_crc32(0, Z_NULL, 0);
+	*CRC = ri.zlib_crc32(*CRC, *(byte**)buffer-4, 4);
+	*crcPtr = *buffer;
+}
+
+/*
+ * Write CRC to buffer.
+ */
+
+void WriteCRC(void **buffer, void **crcPtr, PNG_ChunkCRC CRC)
+{
+	int size = (intptr_t)*buffer-(intptr_t)*crcPtr;
+
+	/*
+	 *  Update CRC
+	 */
+	if (size > 0)
+		CRC = ri.zlib_crc32(CRC, *crcPtr, size);
+
+	/*
+	 *  Write CRC
+	 */
+	CRC = BigLong(CRC);
+	WriteToBuffer(buffer, &CRC, PNG_ChunkCRC_Size);
+}
+
+/*
+ * The PNG saver
+ */
+
+void RE_SavePNG(const char *filename, int width, int height, byte *data, int padding) {
+	void					*pngData;
+	size_t					pngSize;
+	void					*buffer;
+	struct PNG_Chunk_IHDR	IHDR;
+	PNG_ChunkCRC			CRC;
+	void					*crcPtr;
+	int						numtEXt = 0;
+	#define					NUMTEXT 7
+	struct
+	{
+		char key[80]; // PNG limits to 79+'\0'.
+		char text[256]; // PNG allows any length.
+	} tEXt[NUMTEXT];
+	int						i;
+	uint8_t					*imageData;
+	uint32_t				imageLength;
+	Bytef					*compressedData = NULL;
+	uLongf					compressedDataLength;
+
+	/*
+	 *  Create the png image data from buffer.
+	 */
+
+	imageLength = (width*PNG_NumColourComponents_True + 1) * height;
+	imageData = ri.Malloc(imageLength);
+
+	EncodeImageNonInterlaced8True(width, height, data, 3, padding, imageData);
+
+	/*
+	 *  Compress the png image data.
+	 */
+
+	compressedDataLength = imageLength * 1.01f + 12;
+	compressedData = ri.Malloc(compressedDataLength);
+
+	if (ri.zlib_compress(compressedData, &compressedDataLength, imageData, imageLength) != Z_OK) {
+		ri.Free(compressedData);
+		ri.Free(imageData);
+		ri.Printf(PRINT_WARNING, "RE_SavePNG: Failed to compress image data.\n");
+		return;
+	}
+
+	/*
+	 *  Setup text data.
+	 */
+
+	Q_strncpyz(tEXt[numtEXt].key, "Title", sizeof (tEXt[numtEXt].key));
+	ri.Cvar_VariableStringBuffer("version", tEXt[numtEXt].text, sizeof (tEXt[numtEXt].text));
+	numtEXt++;
+	Q_strncpyz(tEXt[numtEXt].key, "Author", sizeof (tEXt[numtEXt].key));
+	ri.Cvar_VariableStringBuffer("username", tEXt[numtEXt].text, sizeof (tEXt[numtEXt].text));
```

Comment:
While I don't mind much about my playername (e.g. `^1That^6Guy`) to show up as metadata in screenshots, I'd rather not have my username (e.g. `John Doe`) recorded somewhere. Especially since the average user might not know about this somewhat hidden metadata or how to remove/edit it.